src/lenses/CrvRefundLens.sol

```diff
@@ -0,0 +1,153 @@
+    uint256 constant BPS_PRECISION = 1e4;
+
+    constructor() {
+        marketLens = IMarketLens(LENS_ADDR);
+        gaugeController = IGaugeController(CURVE_GAUGE_CONTROLLER_ADDR);
+        voteEscrowedCrv = IVoteEscrowedCrv(VE_CRV_ADDR);
+        yBribeContract = IYBribeV3(YBRIBE_V3_ADDR);
+    }
+
+    function getUserBorrowAmounts(ICauldronV4[] calldata cauldrons, address user) public view returns (uint256[] memory borrows) {
+        borrows = new uint256[](cauldrons.length);
+        for (uint256 i = 0; i < cauldrons.length; i++) {
+            borrows[i] = marketLens.getUserBorrow(address(cauldrons[i]), user);
+        }
+    }
+
+    function getVoterMimGaugeVotes(address votingAddress) public view returns (uint256) {
+        uint256 voterVeCrv = getVoterVeCrv(votingAddress);
+        uint256 mimGaugePower = getVoterMimGaugePower(votingAddress);
+        return (voterVeCrv * mimGaugePower) / 1e4;
+    }
+
+    function getVoterVeCrv(address votingAddress) public view returns (uint256) {
+        return voteEscrowedCrv.balanceOf(votingAddress);
+    }
+
+    // Get MIM gauge power of a voter. Power represents % of veCrv balance applied to gauge.
+    function getVoterMimGaugePower(address votingAddress) public view returns (uint256) {
+        (, uint256 power, ) = gaugeController.vote_user_slopes(votingAddress, CURVE_MIM_GAUGE_ADDR);
+        return power;
+    }
+
+    function getSpellPrice() public view returns (uint256) {
+        IOracle oracle = IOracle(SPELL_ORACLE_ADDR);
+        bytes memory data = abi.encodePacked(uint256(0));
+        return PRECISION ** 2 / oracle.peekSpot(data);
+    }
+
+    function getTotalMimGaugeVotes() public view returns (uint256) {
+        return gaugeController.get_gauge_weight(CURVE_MIM_GAUGE_ADDR);
+    }
+
+    function getWeeklySpellBribes() public view returns (uint256) {
+        uint256 rewardsPerGauge = yBribeContract.reward_per_gauge(CURVE_MIM_GAUGE_ADDR, SPELL_ADDR);
+        uint256 claimsPerGauge = yBribeContract.claims_per_gauge(CURVE_MIM_GAUGE_ADDR, SPELL_ADDR);
+        return rewardsPerGauge - claimsPerGauge;
+    }
+
+    function getVoterSpellBribes(address votingAddress) public view returns (uint256) {
+        uint256 totalMimGaugeVotes = getTotalMimGaugeVotes();
+        uint256 voterMimGaugeVotes = getVoterMimGaugeVotes(votingAddress);
+        uint256 weeklySpellBribes = getWeeklySpellBribes();
+
+        // Pro-rate total SPELL bribes by voter's share of gauge votes
+        return (weeklySpellBribes * voterMimGaugeVotes) / totalMimGaugeVotes;
+    }
+
+    function getVoterSpellBribesUsd(address votingAddress) public view returns (uint256) {
+        return (getVoterSpellBribes(votingAddress) * getSpellPrice()) / PRECISION;
+    }
+
+    function getRefundInfo(ICauldronV4[] calldata cauldrons, address user, address votingAddress) public view returns (RefundInfo memory) {
+        address[] memory cauldronContracts = new address[](cauldrons.length);
```

Comment:
you can try something like this instead of the loop

```
address[] memory cauldronContracts;
assembly {
    cauldronContracts := cauldrons
}
refundInfo.cauldrons = cauldronContracts;
```

Reply:
I get this error when I try that:

`Call data elements cannot be accessed directly. Use ".offset" and ".length" to access the calldata offset and length of this array and then use "calldatacopy".solidity(1397)`